libraries/ms-helper/src/can_uart.c

```diff
@@ -0,0 +1,108 @@
+#include "can_uart.h"
+#include <string.h>
+#include "log.h"
+
+// CTX in ASCII
+#define CAN_UART_TX_MARKER 0x585443
+// CRX in ASCII
+#define CAN_UART_RX_MARKER 0x585243
+
+// Defined protocol: Header (u32), ID (u32), Data (u64), newline (u8)
+// Bits  | Field
+// ------|--------------------
+// 0:23  | marker (CTX || CRX)
+// 24    | extended
+// 25    | rtr
+// 26-27 | reserved
+// 28:31 | dlc
+
+#define CAN_UART_BUILD_HEADER(marker, extended, rtr, dlc)             \
+  (((uint32_t)(marker)&0xFFFFFF) | ((uint32_t)(extended)&0x1) << 24 | \
+   ((uint32_t)(rtr)&0x1) << 25 | ((uint32_t)(dlc)&0xF) << 28)
+#define CAN_UART_EXTRACT_HEADER(header, marker, extended, rtr, dlc) \
+  ({                                                                \
+    uint32_t _header = header;                                      \
+    *(marker) = _header & 0xFFFFFF;                                 \
+    *(extended) = (_header >> 24) & 0x1;                            \
+    *(rtr) = (_header >> 25) & 0x1;                                 \
+    *(dlc) = (_header >> 28) & 0xF;                                 \
+    true;                                                           \
+  })
+
+typedef struct CanUartPacket {
+  uint32_t header;
+  uint32_t id;
+  uint64_t data;
+} CanUartPacket;
+
+static void prv_rx_uart(const uint8_t *rx_arr, size_t len, void *context) {
+  CanUart *can_uart = context;
+
+  CanUartPacket packet;
+  if (len < sizeof(packet)) {
+    // no way this is valid
```

Comment:
Should we report a status here?

Reply:
No, this is an internal callback that's run from the UART ISR. I don't think there's much value in returning a status, especially since it's very possible for something to get messed up over UART.

We should probably figure out why we're seeing so many errors tbh. I might have a race condition or something.